library/src/main/java/org/apache/apex/malhar/lib/fs/FSRecordReaderModule.java

```diff
@@ -0,0 +1,332 @@
+ * input directory<br/>
+ * 4. recursive: if scan recursively input directories<br/>
+ * 5. blockSize: block size used to read input blocks of file<br/>
+ * 6. readersCount: count of readers to read input file<br/>
+ * 7. sequentialFileRead: If emit file blocks in sequence?<br/>
+ * 8. blocksThreshold: number of blocks emitted per window
+ */
+@org.apache.hadoop.classification.InterfaceStability.Evolving
+public class FSRecordReaderModule implements Module
+{
+  @NotNull
+  @Size(min = 1)
+  private String files;
+  private String filePatternRegularExp;
+  @Min(0)
+  private long scanIntervalMillis;
+  private boolean recursive = true;
+  private boolean sequentialFileRead = false;
+  private int readersCount;
+  @Min(1)
+  protected int blocksThreshold;
+
+  public final transient ProxyOutputPort<byte[]> records = new ProxyOutputPort<byte[]>();
+
+  /**
+   * Criteria for record split
+   */
+  private RECORD_READER_MODE mode;
+
+  /**
+   * Length for fixed width record
+   */
+  private int recordLength;
+
+  public FileSplitterInput createFileSplitter()
+  {
+    return new FileSplitterInput();
+  }
+
+  public FSRecordReader createBlockReader()
+  {
+    FSRecordReader recordReader = new FSRecordReader();
+    recordReader.setMode(mode);
+    recordReader.setRecordLength(recordLength);
+
+    return recordReader;
+  }
+
+  @Override
+  public void populateDAG(DAG dag, Configuration configuration)
+  {
+    FileSplitterInput fileSplitter = dag.addOperator("FileSplitter", createFileSplitter());
+    FSRecordReader recordReader = dag.addOperator("BlockReader", createBlockReader());
+
+    dag.addStream("BlockMetadata", fileSplitter.blocksMetadataOutput, recordReader.blocksMetadataInput);
+
+    if (sequentialFileRead) {
+      dag.setInputPortAttribute(recordReader.blocksMetadataInput, Context.PortContext.STREAM_CODEC,
+          new SequentialFileBlockMetadataCodec());
+    }
+
+    FileSplitterInput.TimeBasedDirectoryScanner fileScanner = fileSplitter.getScanner();
+    fileScanner.setFiles(files);
+    if (scanIntervalMillis != 0) {
+      fileScanner.setScanIntervalMillis(scanIntervalMillis);
+    }
+    fileScanner.setRecursive(recursive);
+    if (filePatternRegularExp != null) {
+      fileSplitter.getScanner().setFilePatternRegularExp(filePatternRegularExp);
+    }
+
+    recordReader.setBasePath(files);
+    if (readersCount != 0) {
+      dag.setAttribute(recordReader, Context.OperatorContext.PARTITIONER,
+          new StatelessPartitioner<FSSliceReader>(readersCount));
+    }
+    fileSplitter.setBlocksThreshold(blocksThreshold);
+    records.set(recordReader.records);
+  }
+
+  /**
+   * A comma separated list of directories to scan. If the path is not fully
+   * qualified the default file system is used. A fully qualified path can be
+   * provided to scan directories in other filesystems.
+   *
+   * @param files
+   *          files
+   */
+  public void setFiles(String files)
+  {
+    this.files = files;
```

Comment:
Shouldn't the field be named "directories" if we expect it to be a list of directories ?